### PR TITLE
Unfurl integration

### PR DESCRIPTION
**Pull request by jkppr**
Integrating the [dfir-unfurl](https://github.com/obsidianforensics/unfurl) project into timesketch via the context links. This feature will allow for using unful on every URL in a timesketch `url` attribute (configurable). Having the unfurl function included in Timesketch allows for easy analysis and understanding of URLs encountered during the investigation.

image

This PR does:

* Add the concept of "build-in modules" to the context links.
* Add `dfir-unfurl` as a requirement for timesketch.
* Add an API endpoint to handle unfurl requests and return the graph in a json format.
* Add the API call to the frontend API client. (+ fix some string formatting)
* Add an UnfurlDialog component that will render the unfurl graph using the cytoscape js lib
* Adjust the EventDetail component to trigger the unfurl dialog for defined attributes.

Open tasks:

- [x] Fine tune the cytoscape graph:
  - [x] Center graph on loading
  - [x] Dynamically scale the dialog?
  - [x] Tweak the layout and style to look like the original unfurl graph.
- [x] Add mouseover in the graph to show unfurl context on edges and nodes.
- [x] Update the context links to support modules
- [x] Update the `tsctl validate-context-links-conf`
- [x] Update unit tests

### CODE REVIEW

File: timesketch/frontend-ng/src/components/Explore/UnfurlDialog.vue [hunk 1]
```diff
@@ -0,0 +1,174 @@
+<!--
+Copyright 2022 Google Inc. All rights reserved.
```
Comment (berggren) [inline timesketch/frontend-ng/src/components/Explore/UnfurlDialog.vue:2]:
```suggestion
Copyright 2023 Google Inc. All rights reserved.
```

**Comment by berggren** [comment]
AHey @jkppr is this still draft or is it ready for review?

**Comment by jkppr** [comment]
@berggren it is still in draft. There is some fine tuning around the graph and the setup to be done.

**Comment by berggren** [comment]
> @berggren it is still in draft. There is still some fine tuning around the graph and the setup to be done.

Ack, thanks for the update. Let me know when it is ready for review.

Also, @obsidianforensics for awareness, unfurl is being integrated to TS :)

**Comment by berggren** [comment]
@obsidianforensics One idea, could we use the official Unfurl logo in the TS UI? That would be nice :)

**Comment by obsidianforensics** [comment]
Sorry, didn't see the comment until now :/ 

This is amazing! And yes, please use the official logos wherever you'd like. https://github.com/obsidianforensics/unfurl/tree/main/unfurl/static has a number of different types - I'd probably recommend the square `u` one rather than the recentagle `unfurl` one, but feel free to use any.

File: test_tools/test_events/mock_context_links.yaml [hunk 2]
```diff
@@ -1,20 +1,37 @@
-## Mock configuration file for testing the contrext links API endpoint!
-lookupone:
-  short_name: 'LookupOne'
-  match_fields: ['hash']
-  validation_regex: '/^[0-9a-f]{40}$|^[0-9a-f]{32}$/i'
-  context_link: 'https://lookupone.local/q=<ATTR_VALUE>'
-  redirect_warning: TRUE
+# Mock configuration file for testing the contrext links API endpoint!
```
Comment (berggren) [inline test_tools/test_events/mock_context_links.yaml:1]:
```suggestion
# Mock configuration file for testing the context links API endpoint!
```

File: timesketch/frontend-ng/src/components/Explore/UnfurlDialog.vue [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+<template>
+  <v-card>
+    <v-card-title>
+      <v-img src="/unfurl-logo.png" max-height="50" max-width="200" contain class="mb-2"></v-img>
```
Comment (berggren) [inline timesketch/frontend-ng/src/components/Explore/UnfurlDialog.vue:16]: IMO it is a bit big, can we scale it down?

File: timesketch/tsctl.py [hunk 4]
```diff
@@ -645,12 +661,27 @@ def validate_context_links_conf(path):
...
-        try:
-            validate(instance=context_link_config[entry], schema=schema)
-            print(f'=> OK: "{entry}"')
-        except (ValidationError, SchemaError) as err:
-            print(f'=> ERROR: "{entry}" >> {err}\n')
+    if context_link_config["hardcoded_modules"]:
+        for entry in context_link_config["hardcoded_modules"]:
+            try:
+                validate(
+                    instance=context_link_config["hardcoded_modules"][entry],
```
Comment (berggren) [inline timesketch/tsctl.py:668]: Can you explain what "hardcoded_modules" are and why we need them?

File: timesketch/frontend-ng/public/unfurl-logo.png
Comment (berggren) [inline timesketch/frontend-ng/public/unfurl-logo.png:1]: This image does not look good in dark mode:

Screenshot 2023-10-24 at 14 24 01

Can we only use the tree branch perhaps?

**Review by berggren — CHANGES_REQUESTED** [review]

File: timesketch/frontend-ng/public/unfurl-logo.png
Comment (obsidianforensics) [reply timesketch/frontend-ng/public/unfurl-logo.png:1]: There's also https://github.com/obsidianforensics/unfurl/blob/main/unfurl/static/unfurl_dark.png, if you want the same logo but tweaked for a black background

File: timesketch/frontend-ng/src/components/Explore/UnfurlDialog.vue (hunk 3 above)
Comment (jkppr) [reply timesketch/frontend-ng/src/components/Explore/UnfurlDialog.vue]: fixed

File: timesketch/frontend-ng/public/unfurl-logo.png
Comment (jkppr) [reply timesketch/frontend-ng/public/unfurl-logo.png:1]: Thanks for the dark logo! 
I went with the 32x32 version for now to make it a bit smaller:
image

And dark theme: 
image


@obsidianforensics does this work with you? Or would you prefer us using the one linked above?

File: timesketch/tsctl.py (hunk 4 above)
Comment (jkppr) [reply timesketch/tsctl.py:668]: See the explanation in the [context_links.yaml](https://github.com/google/timesketch/pull/2897/files#diff-269e304c1e3c351af85fde89dec9ea1c38dc42663291d29bf783e4c27ee1a19cR13) file. 

TL;DR: `hardcoded_moduled` are context links that are predefined in timesketch. E.g. unfurl and xml formatter. Having them as part of the configuration allows for TS operators to deactivate them if they don't want them and also easily define what fields they should trigger on.

File: timesketch/frontend-ng/public/unfurl-logo.png
Comment (jkppr) [reply timesketch/frontend-ng/public/unfurl-logo.png:1]: Thanks for the dark logo, I added it to the dialog:

image

**Review by berggren — APPROVED** [review]
